### PR TITLE
fix #38749, postgresql GRANT user's permission after createDatabase, …

### LINKED ISSUE #38749: [Bug]: fresh install with postgresql, occ maintenance:install error
### ⚠️ This issue respects the following points: ⚠️

- [X] This is a **bug**, not a question or a configuration/webserver/proxy issue.
- [X] This issue is **not** already reported on [Github](https://github.com/nextcloud/server/issues?q=is%3Aopen+is%3Aissue+label%3Abug) OR [Nextcloud Community Forum](https://help.nextcloud.com/) _(I've searched it)_.
- [X] Nextcloud Server **is** up to date. See [Maintenance and Release Schedule](https://github.com/nextcloud/server/wiki/Maintenance-and-Release-Schedule) for supported versions.
- [X] I agree to follow Nextcloud's [Code of Conduct](https://nextcloud.com/contribute/code-of-conduct/).

### Bug description

Fresh install nextcloud & postgresql, when `occ maintenance:install`, will post an error something like:
```
PostgreSQL username and/or password not valid
 -> You need to enter details of an existing account.
Trace: #0 /var/www/lib/private/Setup.php(353): OC\Setup\PostgreSQL->setupDatabase()
#1 /var/www/core/Command/Maintenance/Install.php(104): OC\Setup->install()
#2 /var/www/3rdparty/symfony/console/Command/Command.php(255): OC\Core\Command\Maintenance\Install->execute()
#3 /var/www/3rdparty/symfony/console/Application.php(1009): Symfony\Component\Console\Command\Command->run()
#4 /var/www/3rdparty/symfony/console/Application.php(273): Symfony\Component\Console\Application->doRunCommand()
#5 /var/www/3rdparty/symfony/console/Application.php(149): Symfony\Component\Console\Application->doRun()
#6 /var/www/lib/private/Console/Application.php(215): Symfony\Component\Console\Application->run()
#7 /var/www/console.php(100): OC\Console\Application->run()
#8 /var/www/occ(11): require_once('...')
#9 {main}

Previous: Doctrine\DBAL\Exception: Failed to connect to the database: An exception occurred in the driver: SQLSTATE[08006] [7] connection to server at "postgres" (192.168.16.6), port 5432 failed: FATAL:  database "nextcloud" does not exist
Trace: #0 /var/www/lib/private/Setup/PostgreSQL.php(112): OC\DB\Connection->connect()
#1 /var/www/lib/private/Setup.php(353): OC\Setup\PostgreSQL->setupDatabase()
#2 /var/www/core/Command/Maintenance/Install.php(104): OC\Setup->install()
#3 /var/www/3rdparty/symfony/console/Command/Command.php(255): OC\Core\Command\Maintenance\Install->execute()
#4 /var/www/3rdparty/symfony/console/Application.php(1009): Symfony\Component\Console\Command\Command->run()
#5 /var/www/3rdparty/symfony/console/Application.php(273): Symfony\Component\Console\Application->doRunCommand()
#6 /var/www/3rdparty/symfony/console/Application.php(149): Symfony\Component\Console\Application->doRun()
#7 /var/www/lib/private/Console/Application.php(215): Symfony\Component\Console\Application->run()
#8 /var/www/console.php(100): OC\Console\Application->run()
#9 /var/www/occ(11): require_once('...')
#10 {main}
```

### Steps to reproduce

1. fresh nextcloud
2. fresh postgresql with superuser `postgres` & thepassword
3. occ maintenance:install --database pgsql --database-host localhost:5432 --database-name nextcloud --database-user postgres --database-pass thepassword --admin-user admin --admin-pass adminpassword--data-dir /path/to/data/

### Expected behavior

`Nextcloud was successfully installed`

### Installation method

Community Docker image

### Nextcloud Server version

26

### Operating system

Debian/Ubuntu

### PHP engine version

PHP 8.1

### Web server

Nginx

### Database engine version

PostgreSQL

### Is this bug present after an update or on a fresh install?

Fresh Nextcloud Server install

### Are you using the Nextcloud Server Encryption module?

Encryption is Disabled

### What user-backends are you using?

- [X] Default user-backend _(database)_
- [ ] LDAP/ Active Directory
- [ ] SSO - SAML
- [ ] Other

### Configuration report

_No response_

### List of activated Apps

_No response_

### Nextcloud Signing status

_No response_

### Nextcloud Logs

_No response_

### Additional info

This bug is related to this pull request. https://github.com/nextcloud/server/pull/34645
Because when execute `$connectionMainDatabase->executeQuery('GRANT CREATE ON SCHEMA public TO ' . addslashes($this->dbUser));` the postgres database may not create.

https://github.com/nextcloud/server/blob/eb7b682dc08f522212a86636994abc90ea926616/lib/private/Setup/PostgreSQL.php#L82

Same issue: https://github.com/nextcloud/server/commit/eb7b682dc08f522212a86636994abc90ea926616#commitcomment-95275498

### PR DESCRIPTION
**Pull request by whlsxl**
…ensure database exist when GRANT

* Resolves: #38749

### CODE REVIEW

File: lib/private/Setup/PostgreSQL.php [hunk 1]
```diff
@@ -94,6 +85,18 @@ public function setupDatabase($username) {
 			$this->createDatabase($connection);
 			// the connection to dbname=postgres is not needed anymore
 			$connection->close();
+
+			if ($canCreateRoles) {
```
Comment (kesselb) [inline lib/private/Setup/PostgreSQL.php]:
```suggestion
			if ($this->tryCreateDbUser && $canCreateRoles) {
```

It's possible to disable the user creation. 
If you do, the variables `$canCreateRoles` and `$connectionMainDatabase` are uninitialized.

File: lib/private/Setup/PostgreSQL.php [hunk 2]
```diff
@@ -61,7 +61,8 @@ public function setupDatabase($username) {
 				} catch (DatabaseException $e) {
 					$canCreateRoles = false;
 				}
-
+				
+				$connectionMainDatabase;
```
Comment (kesselb) [inline lib/private/Setup/PostgreSQL.php]:
```suggestion
```

Either initialize it or remove.

**Review by kesselb — COMMENTED** [review]
Thank you very much for your pull :+1: 

Could you please provide us some more information about your change? 

I'm not familiar with PostgreSQL. Grant create does not work, before creating the database because the schema public does not exist?

**Comment by kesselb** [comment]
I think we should remove the code to create database tables and users.

**Comment by whlsxl** [comment]
> Thank you very much for your pull 👍
> 
> Could you please provide us some more information about your change?
> 
> I'm not familiar with PostgreSQL. Grant create does not work, before creating the database because the schema public does not exist?

Sorry, not familiar with PHP.
This is the `docker-compose.yml` to reproduce this issue.

```
version: "3.5"

services:
  postgres:
    container_name: postgres
    image: postgres:15.3-alpine
    restart: unless-stopped
    environment:
      - "POSTGRES_USER=postgres"
      - "POSTGRES_PASSWORD=U8PNOtHPefQ"
      - "POSTGRES_HOST_AUTH_METHOD=trust"
  nextcloud:
    container_name: nextcloud
    image: nextcloud:26.0.2-fpm-alpine
    restart: unless-stopped
```

After container up, enter the container `docker exec -it nextcloud sh` & install nextcloud with 
`php occ maintenance:install --database pgsql --database-host postgres:5432 --database-name nextcloud --database-user postgres --database-pass U8PNOtHPefQ --admin-user admin --admin-pass U8PNOtHPefQ --data-dir /var/www/html/data/`

got error:
`PostgreSQL username and/or password not valid`

**Comment by whlsxl** [comment]
I traced this bug.
The pull request https://github.com/nextcloud/server/pull/34645 is to fix the https://github.com/nextcloud/server/issues/34617, but I tried many ways, can't reproduce the problem.

> I think we should remove the code to create database tables and users.

In my opinion, create users & database & tables are very good functions, because most users are newbies, they don't have the ability to manage the database.

But this function should be refactor,  for example https://github.com/nextcloud/server/blob/63bf207ca7a18dd50ce3aeaea42e53f4ee400fc0/lib/private/Setup/PostgreSQL.php#L122-L143  if the database exists, the superuser should resign the `owner` to nextcloud database user and copy all the permission nextcloud database user.

**Comment by kesselb** [comment]
> In my opinion, create users & database & tables are very good functions, because most users are newbies, they don't have the ability to manage the database.

I would prefer that newbies learn how to create a database and database user ;)

We have automated tests for PostgreSQL. 

https://github.com/nextcloud/.github/blob/dc0440094e086858cb960e0054a29ca565c5fbbd/workflow-templates/phpunit-pgsql.yml#L49-L52

The notable difference: 
- PostgreSQL 14 vs 15
- POSTGRES_DB (we ask the container to create a database named nextcloud)

That means in our automated tests, we do not run the code to create a database table. This code path is apparently broken. 

```
Previous: Doctrine\DBAL\Exception: Failed to connect to the database: An exception occurred in the driver: SQLSTATE[08006] [7] FATAL:  database "nextcloud" does not exist
Trace: #0 /var/www/html/lib/private/Setup/PostgreSQL.php(114): OC\DB\Connection->connect()
#1 /var/www/html/lib/private/Setup.php(353): OC\Setup\PostgreSQL->setupDatabase('admin')
```

The stack trace is a bit misleading. 

https://github.com/nextcloud/server/blob/eb7b682dc08f522212a86636994abc90ea926616/lib/private/Setup/PostgreSQL.php#L65

Here is the problem. `connect()` uses `this->dbName` by default. That means we tell doctrine/pdo to connect to a non-existing database. 

```
$connectionMainDatabase = $this->connect([
    'dbname' => 'postgres'
]);
```

We need to overwrite the dbname (like for `$connection` a few lines above) and hope that the postgres table will be there.

**Comment by whlsxl** [comment]
This bug still exists, wish someone could review it.

**Review by icewind1991 — APPROVED** [review]

**Comment by susnux** [comment]
Can you please adjust your commits to comply with conventional commits? Thank you :)

**Comment by whlsxl** [comment]
> Can you please adjust your commits to comply with conventional commits? Thank you :)

Are you referring to the commit message? can you tell me more specifically?

**Comment by susnux** [comment]
> Are you referring to the commit message? can you tell me more specifically?

Yes the commit message, we try to follow the conventional commits spec: https://www.conventionalcommits.org/

Some examples:
* `fix: Something that was wrong`
* `fix(DB): Something that was wrong in database`
* `feat: Something new`

**Review by skjnldsv — APPROVED** [review]

**Comment by skjnldsv** [comment]
Rebased and reviewed!
this issue was a bit old and staled, but this should be simple enough to get merged :crossed_fingers: 

Thanks for the hard work @whlsxl :muscle:

**Pull request merged**